### PR TITLE
fix(wallet): error message on scanning different key pair QR

### LINKED ISSUE #20548: Toast error messages issues when scanning a "missing key pair" QR code
# Bug Report

## Problem 

Two unhappy flows that were initially reported in https://github.com/status-im/status-mobile/pull/20464#issuecomment-2186348441 are described here, but further investigation may be needed on how we should handle toast errors in all possible cases.

**Designs:** https://www.figma.com/design/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?node-id=232-33031&t=hioBwxRjpwpVBl41-0

### 1. Wrong error text when scanning a QR from a wrong key pair when attempting to import a missing key pair

**Steps:**
1. Create a profile on Device1
2. Sync Device2 with Device1
3. Create a new key pair on Device1
4. On Device1: open `Profile` > `Wallet` > `Key pairs and accounts` > press three dots on some key pair, other than needed one (e.g. on the default key pair) > `Show encrypted QR of key pairs on device`
5. On Device2: press three dots on the missing key pair > `Import by scanning encrypted QR` > Scan the wrong QR

**Expected error:** `This is not the QR you are looking for` (pls correct me if I'm wrong here)
**Actual error:** 

Screenshot 2024-06-24 at 12 59 25

### 2. Wrong error text when scanning an already used QR

This might be an edge case, but I honestly don't know how else to reproduce the "QR code has already been used" error, so I tried this scenario

**Steps:**
1. Create a profile on Device1
2. Sync Device2 and Device3 with Device1
3. Create a new key pair on Device1
4. Scan the required QR code for the missing key pair on Device2
5. Scan **_the same_** QR code for the missing key pair on Device3

**Expected error:** `QR code has already been used`
**Actual error:**

Screenshot 2024-06-24 at 13 25 35

### Additional Information

- Status version: https://github.com/status-im/status-mobile/pull/20464 
- Operating System: Android, iOS

cc @smohamedjavid

### PR DESCRIPTION
**Pull request by smohamedjavid**
fixes #20548

### Summary

This PR 
- update the connection string validation method to use the method from status-go
- updates the error message (`This QR does not contain {{name}} key pair`) if the user tries to scan a different key pair QR for importing a missing key pair
- updates the text for exporting an individual key pair

### Testing notes

A quick regression on pairing devices (generating QR and scanning QR) is appreciated as this updates the validation method for the connection string.

### Platforms

- Android
- iOS

#### Prerequisite: A user with 2 paired devices (A & B) and a missing key pair on one device (eg: device B)

On Device A:

- Open Status
- Navigate to `profile > wallet > key pairs and accounts`
- On a key pair which is NOT imported on device B, tap on `Show encrypted QR of key pair`

On Device B:

- Open Status
- Navigate to `profile > wallet > key pairs and accounts`
- Import the missing key pair by scanning the QR on device A
- Verify the toast message is shown if you try to scan a different key pair QR 

status: ready

### CODE REVIEW

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/Utils.kt [hunk 1]
```diff
@@ -153,4 +153,9 @@ class Utils(private val reactContext: ReactApplicationContext) : ReactContextBas
 
         return strArray
     }
+
+    @ReactMethod(isBlockingSynchronousMethod = true)
+    fun validateConnectionString(connectionString: String, callback: Callback) {
+        executeRunnableStatusGoMethod({ Statusgo.validateConnectionString(connectionString) }, callback)
+    }
 }
```
Comment (seanstrom) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/Utils.kt]: Do we need to pass `callback` here if the method is synchronous?

File: modules/react-native-status/ios/RCTStatus/Utils.m [hunk 2]
```diff
@@ -128,4 +128,8 @@ + (void) migrateKeystore:(NSString *)accountData
     return StatusgoToChecksumAddress(address);
 }
 
+RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(validateConnectionString:(NSString *)cs) {
+    return StatusgoValidateConnectionString(cs);
+}
+
```
Comment (seanstrom) [inline modules/react-native-status/ios/RCTStatus/Utils.m:134]: This is a nice example of defining a synchronous function binding 💯

File: src/native_module/core.cljs [hunk 3]
```diff
@@ -533,6 +533,13 @@
    (log/debug "[native-module] validate-mnemonic")
    (.validateMnemonic ^js (utils) mnemonic callback)))
 
+(defn validate-connection-string
+  [connection-string]
+  (log/debug "[native-module] validate-connection-string")
+  (->> connection-string
+       (.validateConnectionString ^js (utils))
+       types/json->clj))
```
Comment (seanstrom) [inline src/native_module/core.cljs:551]: How does this currently execute on Android if we do not pass a callback?

File: src/status_im/contexts/settings/wallet/events.cljs [hunk 4]
```diff
@@ -95,20 +96,34 @@
...
-                                                :text  (i18n/label :t/incorrect-qr-code)}]))}]]}))
+            :on-success (fn [key-uids]
+                          (rf/call-continuation on-success)
+                          (rf/dispatch [:wallet/make-keypairs-accounts-fully-operable key-uids]))
+            :on-fail (fn [error]
+                       (let [error-message (-> error ex-data :error)
+                             incorrect-keypair?
+                             (string/includes?
+                              error-message
+                              "one or more expected keystore files are not found among the sent files")
```
Comment (seanstrom) [inline src/status_im/contexts/settings/wallet/events.cljs]: We may be able to handle this error string inside the effect `:effects.syncing/import-keypairs-keystores`, and then that effect could throw a more descriptive exception. Thoughts?

**Review by seanstrom — COMMENTED** [review]
Looks good 👍 
I left some small comments / questions 🙏

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/Utils.kt (hunk 1 above)
Comment (smohamedjavid) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/Utils.kt]: Good catch! I missed the Android part when moving to the synchronous method 🙏

File: src/status_im/contexts/settings/wallet/events.cljs (hunk 4 above)
Comment (smohamedjavid) [reply src/status_im/contexts/settings/wallet/events.cljs]: I believe it's not needed as we would be displaying a different message for importing a single key pair and for importing all key pairs separately. 

I'm thinking of creating a separate event (as we access app-db for key pair name) to handle the **on-fail** for any cases/flows and keep `wallet/connection-string-for-import-keypair` simple. I haven't created it yet as I need to check the behaviour for all key pairs flow and we can do a clean-up later.

**Review by seanstrom — APPROVED** [review]

**Review by mohsen-ghafouri — APPROVED** [review]

**Comment by mariia-skrypnyk** [comment]
Hi @smohamedjavid !

Thanks for your PR. Your description looks very clear and full filled.

I have several moments to discuss:

1) are these steps correct that after scanning incorrect QR code firstly we enter our password and then we see the toast?
As a user perspective it would be clearer to see the toast before waisting time on password. Let me know your vision.

2) I see unexpected toast while some scanning testing but not sure what is the trigger for it:

Screenshot 2024-07-10 at 12 47 32

I was doing simple scanning to cause your toast "This QR doesn't contain ..." and after second try of scanning the same QR code I get this toast from my screenshot.

**Comment by smohamedjavid** [comment]
@mariia-skrypnyk - Thanks for testing the PR 🙏 

> are these steps correct that after scanning incorrect QR code firstly we enter our password and then we see the toast?
As a user perspective it would be clearer to see the toast before waisting time on password. Let me know your vision.

Actually, we won't know it's the wrong key pair QR until the connection is made with another device and we receive the files and verify what we are supposed to receive. The connection string or the QR code doesn't contain any information about the data that will be transferred (uses local pairing or syncing feature beneath). I agree it's kind of creating a bad UX but we will definitely talk about improving this in the upcoming releases.

> I see unexpected toast while some scanning testing but not sure what is the trigger for it:
> I was doing simple scanning to cause your toast "This QR doesn't contain ..." and after second try of scanning the same QR code I get this toast from my screenshot.

You are right, this error is returned from status-go upon scanning the code a second time. Initially, it was intended to show the error from status-go instead of hiding the errors and making the user feel stuck on the screen. We will look into this later as this error is coming from status-go and it's not due to the changes in this PR.

**Comment by mariia-skrypnyk** [comment]
> I agree it's kind of creating a bad UX but we will definitely talk about improving this in the upcoming releases.

Thanks @smohamedjavid  for understanding my feelings in this logic. It will be really good to talk about it as an improvement, nothing more than that! 🙏

> We will look into this later as this error is coming from status-go and it's not due to the changes in this PR.

 I can create a low prio follow up on this just not to loose this in future.

**Comment by mariia-skrypnyk** [comment]
Hey @smohamedjavid !
Failed e2e ae not related.
You can merge your PR.

**Pull request merged**